### PR TITLE
ARTEMIS-607 Added interceptor support for MQTT protocol.

### PR DESCRIPTION
**Pull request by johnament**
BTW, feel free to cry foul on things that don't look right.  I was trying to follow the patterns in Stomp, but the structure just doesn't line up.  I didn't want to duplicate the logic so I did move into a common base class.  The result though is that the protocol handler knows about its parent protocol handler manager.

### CODE REVIEW

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/mqtt/imported/MQTTTest.java [hunk 1]
```diff
@@ -380,7 +384,8 @@ public void testMQTTPathPatterns() throws Exception {
          Message msg = connection.receive(5, TimeUnit.SECONDS);
          do {
             assertNotNull("RETAINED null " + wildcard, msg);
-            assertTrue("RETAINED prefix " + wildcard, new String(msg.getPayload()).startsWith(RETAINED));
+            String msgPayload = new String(msg.getPayload());
```
Comment (johnament) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/mqtt/imported/MQTTTest.java]: One worrisome thing.  I think there's some interdependence in these tests.  I originally added my interceptor checks in a new test method.  When I did that, this test started failing.  It was receiving the messages created previously.  I suspect the server isn't getting cleaned up 100% correctly, or maybe not at all.

**Comment by jbertram** [comment]
Could you squash these two commits together?

**Comment by clebertsuconic** [comment]
These look good.. I can squash them at the time of the merge.


I will run the whole testsuite just in case. (It takes 2 hours or so now)

**Comment by johnament** [comment]
Up to you.  I won't be able to squash until this evening.

**Comment by johnament** [comment]
Ok, I rebased from master and squashed.

File: docs/user-manual/en/intercepting-operations.md [hunk 2]
```diff
@@ -9,7 +9,9 @@ makes interceptors powerful, but also potentially dangerous.
 
 ## Implementing The Interceptors
 
-An interceptor must implement the `Interceptor interface`:
+All interceptors are protocol specific.
+
+An interceptor for the core protocol must implement the interface `Interceptor`:
```
Comment (jdanekrh) [inline docs/user-manual/en/intercepting-operations.md]: I'd change that to 

    implement the `Interceptor` interface

sounds more like standard Java-speak to me that way.

**Pull request closed without merging**